changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,613 @@
+mathematics behind the birthday problem in probability
+theory. This attack can be used to abuse communication
+between two or more parties. The attack depends on the
+higher likelihood of collisions found between random
+attack attempts and a fixed degree of
+permutations (pigeonholes). With a birthday attack,
+if we have a total of n bits output and therefore
+$2^n$ different outputs, it is possible to find a
+collision of a hash function with 50% chance by
+calculating approximately $sqrt(2^n) = 2^{n/2}$ hashes.
+
+To find a random collision in a SHA512 hash with
+probability 0.5, we would need to calculate
+approximately $2^256$ hashes. Similarly for SHA384,
+we would need to calculate $2^192$ hashes.
+Therefore, SHA512 is more collision resistant.
+
+## Appendix
+
+## A1: Relating Cache Entries (Branch Keys) to KMS Relationships
+
+For some MPL Consumers,
+Cache Entries SHOULD only be served IF
+the request is being served under the same KMS Relationship,
+which is the most secure condition,
+as it prevents a Confused Deputy scenario.
+
+To facilitate this,
+the Cache Entry Identifier SHOULD
+describe the KMS Relationship,
+as well as the Logical Key Store Name,
+Branch Key ID,
+and Branch Key Name.
+
+We can use existing features to accomplish this;
+The Key Store constructor takes an optional [ID](../../framework/branch-key-store.md#keystore-id)
+parameter at construction.
+
+We can use this [ID](../../framework/branch-key-store.md#keystore-id)
+to label the KMS Relationship of a particular Key Store instance.
+
+MPL Consumers who want to garbage collect Key Store instances,
+but still retain the ability to serve cached results,
+MAY then set this [ID](../../framework/branch-key-store.md#keystore-id)
+intelligently.
+
+For example, imagine we have `N` KMS Relationships,
+one of which is `X`,
+which are used when retrieving Branch Keys, `n'`,
+one of which is `x'`.
+
+A common Cryptographic Materials Cache is established
+across all `N` Hierarchical Keyrings used to serve `n'`.
+
+But the Keyrings are routinely garbage collected,
+only the common cache remains "in scope" indefinitely.
+
+Whenever `x'` is needed,
+a new KMS Client is created that respects relationship `X`,
+which is then used to create Key Store Identified as `X`,
+and the Key Store is used to create a Hierarchical Keyring
+that uses the common Cache.
+
+All entries in this common Cache from that hierarchal Keyring
+MUST have a cache identifier that uniquely represents `X`.
+
+When the Hierarchical Keyring for `X` gets garbage collected,
+the cache entries MAY remain.
+
+But to retrieve an entry,
+the MPL Consumer SHOULD recreate
+
+> a KMS Client is created that respects relationship `X`,
+> which is then used to create Key Store Identified as `X`,
+> and the Key Store is used to create a Hierarchical Keyring
+> that uses the common Cache.
+
+If the Cache Entry is still valid
+(the TTL has not expired),
+it can be served without any KMS or DDB requests.
```

Comment:
Do we have any concrete example? Anything that has to do with GC gets tricky and can be source of hard to debug bugs.